File: provider-rustore/src/main/java/io/appmetrica/analytics/push/provider/rustore/impl/BasePushServiceController.java
package io.appmetrica.analytics.push.provider.rustore.impl;

import android.app.Application;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.VisibleForTesting;
import io.appmetrica.analytics.push.coreutils.internal.CoreConstants;
import io.appmetrica.analytics.push.coreutils.internal.utils.PLog;
import io.appmetrica.analytics.push.coreutils.internal.utils.PublicLogger;
import io.appmetrica.analytics.push.coreutils.internal.utils.TrackersHub;
import io.appmetrica.analytics.push.provider.api.PushServiceController;
import io.appmetrica.analytics.push.provider.api.PushServiceExecutionRestrictions;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;
import ru.rustore.sdk.core.tasks.Task;
import ru.rustore.sdk.pushclient.RuStorePushClient;

public class BasePushServiceController implements PushServiceController {

    private static final long DEFAULT_TOKEN_TIMEOUT = 10;
    private static final TimeUnit DEFAULT_TOKEN_TIMEOUT_TIMEUNIT = TimeUnit.SECONDS;
    private static final Long MAX_TASK_EXECUTION_DURATION_SECONDS = 20L;

    @NonNull
    private final Application application;
    @NonNull
    private final IdentifierExtractor identifierExtractor;

    public BasePushServiceController(@NonNull final Application application) {
        this(application, new DefaultIdentifierFromMetaDataExtractor(application));
    }

    @VisibleForTesting
    BasePushServiceController(
        @NonNull final Application application,
        @NonNull final IdentifierExtractor identifierExtractor
    ) {
        this.application = application;
        this.identifierExtractor = identifierExtractor;
    }

    @Override
    public boolean register() {
        PLog.d("Register in RuStore");
        try {
            RuStorePushClient.INSTANCE.init(application, getIdentifier().projectId);
            return true;
        } catch (Throwable ignore) {
            return false;
        }
    }

    @Nullable
    @Override
    public String getToken() {
        TokenResult tokenResult = getTokenOnce();
        if (tokenResult.isSuccessful()) {
            return tokenResult.token;
        } else {
            PublicLogger.e(tokenResult.exception, "Failed to get token, will retry once");
            tokenResult = getTokenOnce();
            if (tokenResult.isSuccessful()) {
                return tokenResult.token;
            } else {
                PublicLogger.e(tokenResult.exception, "Failed to get token after retry");
                TrackersHub.getInstance().reportError("Attempt to get push token failed", tokenResult.exception);
                return null;
            }
        }
    }

    @NonNull
    private TokenResult getTokenOnce() {
        try {
            final CountDownLatch countDownLatch = new CountDownLatch(1);
            final TokenOnCompleteListener listener = new TokenOnCompleteListener(countDownLatch);
            final Task<String> tokenTask = RuStorePushClient.INSTANCE.getToken();
            tokenTask.addOnSuccessListener(listener);
            tokenTask.addOnFailureListener(listener);
            if (!countDownLatch.await(DEFAULT_TOKEN_TIMEOUT, DEFAULT_TOKEN_TIMEOUT_TIMEUNIT)) {
                throw new TimeoutException("token retrieval timeout");
            }
            TokenResult result = listener.getTokenResult();
            if (result == null) {
                result = new TokenResult(null, new IllegalStateException("token retrieval timeout"));
            }
            return result;
        } catch (Throwable e) {
            return new TokenResult(null, e);
        }
    }

    @NonNull
    @Override
    public String getTransportId() {
        return CoreConstants.Transport.RUSTORE;
    }

    @NonNull
    @Override
    public PushServiceExecutionRestrictions getExecutionRestrictions() {
        return new PushServiceExecutionRestrictions() {
            @Override
            public Long getMaxTaskExecutionDurationSeconds() {
                return MAX_TASK_EXECUTION_DURATION_SECONDS;
            }
        };
    }

    @NonNull
    public Identifier getIdentifier() {
        return identifierExtractor.extractIdentifier();
    }

    @NonNull
    public String getExceptionMessage() {
        return identifierExtractor.getExceptionMessage();
    }
}

File: provider-firebase/src/main/java/io/appmetrica/analytics/push/provider/firebase/impl/BasePushServiceController.kt
package io.appmetrica.analytics.push.provider.firebase.impl

import android.annotation.SuppressLint
import android.content.Context
import androidx.annotation.VisibleForTesting
import com.google.android.gms.common.ConnectionResult
import com.google.android.gms.common.GoogleApiAvailability
import com.google.firebase.FirebaseApp
import com.google.firebase.FirebaseOptions
import com.google.firebase.messaging.FirebaseMessaging
import io.appmetrica.analytics.push.coreutils.internal.CoreConstants
import io.appmetrica.analytics.push.coreutils.internal.utils.TrackersHub
import io.appmetrica.analytics.push.logger.internal.DebugLogger
import io.appmetrica.analytics.push.logger.internal.PublicLogger
import io.appmetrica.analytics.push.provider.api.PushServiceController
import io.appmetrica.analytics.push.provider.api.PushServiceExecutionRestrictions
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit
import java.util.concurrent.TimeoutException

private const val DEFAULT_TOKEN_TIMEOUT = 10L
private val DEFAULT_TOKEN_TIMEOUT_TIMEUNIT = TimeUnit.SECONDS

internal open class BasePushServiceController @VisibleForTesting internal constructor(
    val context: Context,
    extractor: IdentifierExtractor
) : PushServiceController {

    private val tag = "[BasePushServiceController]"

    // https://nda.ya.ru/t/qma58mHp76FPa5
    private val maxTaskExecutionDurationSecondsForFirebase = 20L

    val identifier: Identifier by lazy { extractor.extractIdentifier() }
    val exceptionMessage: String by lazy { extractor.exceptionMessage }
    private var firebaseMessaging: FirebaseMessaging? = null

    constructor(context: Context) : this(context, DefaultIdentifierFromResourcesExtractor(context))

    override fun register(): Boolean {
        DebugLogger.info(tag, "Register in Firebase")
        return if (playServicesAvailable()) {
            val firebaseApp = initializeFirebaseApp(identifier.toFirebaseOptions())
            // FirebaseMessaging.getInstance(FirebaseApp) should be public due to documentation but it is package private
            // Found workaround at https://stackoverflow.com/questions/62253691/firebasemessaging-getinstancefirebaseapp-for-secondary-app-supposed-to-be-publ
            firebaseMessaging = firebaseApp.get(FirebaseMessaging::class.java)
            true
        } else {
            PublicLogger.warning("Google play services not available")
            TrackersHub.getInstance().reportEvent("Google play services not available")
            false
        }
    }

    private fun playServicesAvailable(): Boolean {
        return try {
            val apiAvailability = GoogleApiAvailability.getInstance()
            val resultCode = apiAvailability.isGooglePlayServicesAvailable(context)
            resultCode == ConnectionResult.SUCCESS
        } catch (ignored: Throwable) {
            false
        }
    }

    open fun initializeFirebaseApp(firebaseOptions: FirebaseOptions): FirebaseApp {
        try {
            FirebaseApp.initializeApp(context, firebaseOptions)
        } catch (e: Throwable) {
            DebugLogger.error(tag, e, e.message)
        }
        return FirebaseApp.getInstance()
    }

    @SuppressLint("MissingFirebaseInstanceTokenRefresh")
    override fun getToken(): String? {
        val firebaseMessaging = firebaseMessaging
        if (firebaseMessaging == null) {
            TrackersHub.getInstance().reportEvent("Attempt to get push token failed since firebaseMessaging is null")
            return null
        }
        var tokenResult = getToken(firebaseMessaging)
        return if (tokenResult.isSuccess) {
            tokenResult.token
        } else {
            PublicLogger.error(tokenResult.exception, "Failed to get token, will retry once")
            tokenResult = getToken(firebaseMessaging)
            if (tokenResult.isSuccess) {
                tokenResult.token
            } else {
                PublicLogger.error(tokenResult.exception, "Failed to get token after retry")
                TrackersHub.getInstance().reportError("Attempt to get push token failed", tokenResult.exception)
                null
            }
        }
    }

    private fun getToken(firebaseMessaging: FirebaseMessaging): TokenResult {
        return try {
            val countDownLatch = CountDownLatch(1)
            val getTokenTask = firebaseMessaging.token
            getTokenTask.addOnCompleteListener { countDownLatch.countDown() }
            if (!countDownLatch.await(DEFAULT_TOKEN_TIMEOUT, DEFAULT_TOKEN_TIMEOUT_TIMEUNIT)) {
                throw TimeoutException("token retrieval timeout")
            }
            if (getTokenTask.isSuccessful) {
                TokenResult(token = getTokenTask.result)
            } else {
                TokenResult(exception = getTokenTask.exception)
            }
        } catch (e: Throwable) {
            TokenResult(exception = e)
        }
    }

    override fun getTransportId(): String = CoreConstants.Transport.FIREBASE

    override fun getExecutionRestrictions(): PushServiceExecutionRestrictions =
        object : PushServiceExecutionRestrictions() {
            override fun getMaxTaskExecutionDurationSeconds(): Long = maxTaskExecutionDurationSecondsForFirebase
        }

    private class TokenResult(
        val token: String? = null,
        val exception: Throwable? = null
    ) {

        val isSuccess by lazy { exception == null && token != null }
    }

    override fun shouldSendToken(token: String): Boolean = token == getToken()
}
